sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+# Webhooks Beta
+
+
+## PUBLIC
+Authors: @erictune, @caesarxuchao, @enisoc, @lavalamp
+Thanks to: {@smarterclayton, @deads2k, @cheftako, @jpbetz, @mbohlool, @mml, @janetkuo} for comments, data, prior designs, etc.  
+
+
+[TOC]
+
+
+# Summary 
+
+This document proposes a detailed plan for bringing Webhooks to Beta. Highlights include (incomplete, see rest of doc for complete list) :
+
+
+
+*   Adding the ability for webhooks to mutate.  
+*   Bootstrapping
+*   Monitoring
+*   Versioned rather than Internal data sent on hook
+*   Ordering behavior within webhooks, and with other admission phases, is better defined
+
+This plan is compatible with the [original design doc]( https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission_control_extension.md).
+
+
+# Definitions 
+
+**Mutating Webhook**: Webhook that can change a request as well as accept/reject.  
+
+**Non-Mutating Webhook**: Webhook that cannot change request, but can accept or reject.
+
+**Webhook**: encompasses both Mutating Webhook and/or Non-mutating Webhook.
+
+**Validating Webhook**: synonym for Non-Mutating Webhook
+
+**Static Admission Controller**: Compiled-in Admission Controllers, (in plugin/pkg/admission).
+
+**Webhook Host**: a process / binary hosting a webhook.
+
+
+# Development Plan 
+
+Google able to staff development, test, review, and documentation.  Community help welcome, too, esp. Reviewing.
+
+Intent is Beta of Webhooks (**both** kinds) in 1.9.  
+
+Not in scope:
+
+
+
+*   Initializers remains Alpha for 1.9.  (See [Comparison of Webhooks and Initializers](#comparison-of-webhooks-and-initializers) section).  No changes to it.  Will revisit its status post-1.9.
+*   Converting static admission controllers is out of scope (but some investigation has been done, see Moving Built-in Admission Controllers section).
+*   Internal Go interface refactor (e.g. along the lines suggested  #[1137](https://github.com/kubernetes/community/pull/1137)).  Nice-to-have but not gating beta.   
```

Comment:
Along with your other edits promote this to in-scope, we agreed to the mutating/validating split inside the golang code with at least PSP split as a proof of principle.  We're actively working it here https://github.com/kubernetes/kubernetes/pull/54484 and here https://github.com/kubernetes/kubernetes/pull/54689, so I don't expect delivery to be a problem.

Reply:
done